### ISSUE TITLE
extractBufferedData consumes too much memory

### ISSUE THREAD

**Issue by skito** [abbeycode/UnrarKit#70]
Seems that the data in 
`try worker.extractBufferedData(fromFile: fileInfo.filename) { (data, progress) in  ... }`
is not released, because the app consumes about 6GB memory with 3GB multivolume archive. 

Extracting same archive with `extractFiles(to: destinationPath, overwrite: true)` only takes about 20MB in the memory.

Could you check please? Also it would be nice to provide cancelation inside `extractBufferedData` so we don't have to wait 1gb file to be fully extracted before canceling the operation. Maybe thought the `progress` object...

Thanks,
Dimitar

**Comment by abbeycode** [comment]
That's certainly problematic. The reason I wrote the `extractBufferedData` method in the first place was to alleviate memory pressure for large archives. Back when I wrote it I had done profiling to make sure memory doesn't spike. Are you sure you're not retaining references to the data chunks from within your block?

Also, good call on cancellation. Can you create a separate issue for that? Yeah, checking `isCancelled` on the `NSProgress` should do it.

**Comment by skito** [comment]
The code is pretty much your example from the previous issue:
```swift
                // ..................
                // Start extraction on chunks
                try worker.extractBufferedData(fromFile: fileInfo.filename) { (data, progress) in
                    
                    // Append data chunk
                    fileHandle.write(data);
                    totalExtracted += Int64(data.count);
                    
                    // Report progress
                    self._progressTotal = (Double(totalExtracted) / Double(totalSize)) * 100;
                    self.progressChange.Raise(data: (self, ExtractorProgressEventArgs(progressTotal: self._progressTotal)));
                   
                }
                // ..................
```

For reference here's full code of the function which is part of a class:

```swift
private func BExtract(sender: Any, args: Any?) -> Any?
    {
        // INIT Rar Object
        guard let worker = try? URKArchive(path: self._fileArchive) else {
            let error = NSError.init(
                domain: "n/a",
                code: -1,
                userInfo: [NSLocalizedDescriptionKey: "Could't allocate the archive!"]);
            
            print("Extracting error:\n\(error)");
            self.onError.Raise(data: (self, ExtractorErrorEventArgs(error: error)));
            self._isBusy = false;
            return nil;
        };
        
        // Create files list array, to send with OnBefore
        guard let files = try? worker.listFilenames() else {
            let error = NSError.init(
                domain: "n/a",
                code: -2,
                userInfo: [NSLocalizedDescriptionKey: "Couldn't list archive files!"]);
            print("Extracting error:\n\(error)");
            self.onError.Raise(data: (self, ExtractorErrorEventArgs(error: error)));
            self._isBusy = false;
            return nil;
        };
        
        // Create file info list to operate the extraction process with
        guard let fileInfos = try? worker.listFileInfo() else {
            let error = NSError.init(
                domain: "n/a",
                code: -3,
                userInfo: [NSLocalizedDescriptionKey: "Couldn't read the archive files meta!"]);
            print("Extracting error:\n\(error)");
            self.onError.Raise(data: (self, ExtractorErrorEventArgs(error: error)));
            self._isBusy = false;
            return nil;
        };
        
        // Parse destination path
        let destinationPath: String = args as? String ?? "";
        let destinationURL = URL(fileURLWithPath: destinationPath);
        
        // Extract some meta
        let totalSize = worker.uncompressedSize!.int64Value
        var totalExtracted: Int64 = 0;
        
        // Notify listeners
        self.beforeExtraction.Raise(data: (self, BeforeExtractionEventArgs(totalSize: totalSize, files: files)));
        
        // Start extracting files
        for fileInfo in fileInfos {
            
            // Check state
            if self._canceled {
                self.onCanceled.Raise(data: (self, ExtractorProgressEventArgs(progressTotal: self._progressTotal)));
                self._isBusy = false;
                return nil;
            }
            
            do {
                // Prepare the file url and delete any old items
                let fileURL = destinationURL.appendingPathComponent(fileInfo.filename);
                var currentIsDir: ObjCBool = false;
                if FileManager.default.fileExists(atPath: fileURL.path, isDirectory: &currentIsDir) {
                    if !currentIsDir.boolValue {
                        try FileManager.default.removeItem(atPath: fileURL.path);
                    }
                    else {
                        continue;
                    }
                }
                
                // Firstly create the file, otherwise FileHandle will throw error
                // with non-existing file url
                let dirURL = fileURL.deletingLastPathComponent();
                if !FileManager.default.fileExists(atPath: dirURL.path) {
                    try FileManager.default.createDirectory(atPath: dirURL.path, withIntermediateDirectories: true);
                }
                FileManager.default.createFile(atPath: fileURL.path, contents: nil);
                

                // Prepare file handler and make sure to release resources
                // on scope exit, no matter of the result
                let fileHandle = try FileHandle(forWritingTo: fileURL);
                defer {
                    fileHandle.closeFile();
                }

                // Start extraction on chunks
                try worker.extractBufferedData(fromFile: fileInfo.filename) { (data, progress) in
                    
                    // Append data chunk
                    fileHandle.write(data);
                    totalExtracted += Int64(data.count);
                    
                    // Report progress
                    self._progressTotal = (Double(totalExtracted) / Double(totalSize)) * 100;
                    self.progressChange.Raise(data: (self, ExtractorProgressEventArgs(progressTotal: self._progressTotal)));
                   
                }
                
            } catch let error {
                print("Extracting error with file \(fileInfo.filename):\n\(error)");
                self.onError.Raise(data: (self, ExtractorErrorEventArgs(error: error)));
                self._isBusy = false;
                return nil;
            }
        }
        
        // Notify listeners for completition
        self.onFinish.Raise(data: (self));
        
        self._isBusy = false;
        return nil;
    }
```

where `self` is holding some reference properties and event callbacks like

`_canceled`
`_isBusy`
`_progressTotal`
`etc...`

If you want I can paste the full class code to test with. It's not very big and can be used as standalone object, not related to the rest of my project.

I'm creating the cancelation issue.

**Comment by abbeycode** [comment]
Can you try calling wrapping the call to `worker.extractBufferedData` inside an autoreleasepool block?

**Comment by skito** [comment]
It's better with autoreleasepool, but not enough.  Now instead of 6GB memory consumption the archive extraction only takes around 1.5GB. 4-5 peaks by 1.35GB and release after each peak. 

Also I noticed significantly slower performance than `extractFiles` where the archive is extracted for about 1 minute, but with `extractBufferedData` this takes about 5 minutes. The issues is from the beginning, not related to autoreleasepool.

**Comment by abbeycode** [comment]
@skito Did you ever get a better handle on what's going on? If not, I'll take a look when I can.

**Comment by skito** [comment]
Nope. I just decided to go with `extractFiles` as it's much faster. The progress is not such big issue as the cancelation while extracting large file. To handle this and minimize the "Stop" waiting I'm calling `extractFiles` in a separate thread and terminating the whole thread on Cancel request. In my case it's satisfactory.

**Comment by abbeycode** [comment]
@skito Interesting. You're referring to #74?

**Comment by skito** [comment]
Yes. #74

**Comment by skito** [comment]
Yes I'll help to resolve the issue, but I'm currently on a business trip. Will got some time on the end of the next week to assist.

**Comment by skito** [comment]
I'm having hard time to run `xctest` from `large-file-extract-profiling`. I can only run and profile successfully the UnrarExample project which is for iPhone. I've tried to add new OSX Console target to the `UnrarKit` project to run, but can't call the `testExtractBufferedData_VeryLargeSingleFile` function. Could you give me some more detailed guides how to run `xctest` after downloading the branch?

**Comment by abbeycode** [comment]
@skito Sorry it took me a while to respond. Are you using Xcode 9? I'm using 9.3 and I'm able to run the test directly from Xcode. See the screenshot below. Did you select the My Mac device (top of screenshot)?

<img width="661" alt="Xcode 9.3 unit test screenshot" src="https://user-images.githubusercontent.com/1349578/42510768-f0aa5998-841d-11e8-8520-f3971da85682.png">

**Comment by abbeycode** [comment]
@skito Did that help?

**Comment by skito** [comment]
I'm using xCode 9.4.1. This is the point I got to. I'm guessing nothing is happening. I've never did profiling before, so in order to provide you some real data I'll need step-by-step(click this, select this, etc...) guide how to do so with `large-file-extract-profiling` branch.

![screen shot 2018-08-01 at 16 38 37](https://user-images.githubusercontent.com/26840279/43525090-8941fa0c-95a9-11e8-9edb-3cfbd3192602.png)

**Comment by abbeycode** [comment]
Sure, @skito, no problem. I'm no expert in Instruments, but I'll show you how to get where I did for comparison.

1. As mentioned above, copy your archive into the `Tests/Test Data` directory and update the `largeArchiveName` and `archivedFileName` variables for your archive
2. Right-click on the run-test rhombus next to the unit test's name, and choose **Profile "testExtractBufferedData_VeryLargeSingleFile"**

    <img width="429" alt="screen shot 2018-08-09 at 4 39 39 pm" src="https://user-images.githubusercontent.com/1349578/43925595-a4ee911e-9bf5-11e8-9c3a-6d08a28bd66c.png">

3. Choose the **Allocations** instrument when Instruments pops up

    <img width="804" alt="screen shot 2018-08-09 at 4 33 46 pm" src="https://user-images.githubusercontent.com/1349578/43925645-c874669a-9bf5-11e8-8d04-a826fa224394.png">

4. Click the + button in Instruments's top right corner, then choose the **Points of Interest** instrument

    <img width="381" alt="screen shot 2018-08-09 at 4 35 48 pm" src="https://user-images.githubusercontent.com/1349578/43925722-0f3dea1a-9bf6-11e8-9626-cec0e68e0f21.png">

5. Click the red circle ("Record") button in Instruments's top left corner

    <img width="398" alt="screen shot 2018-08-09 at 4 33 55 pm" src="https://user-images.githubusercontent.com/1349578/43925772-3aef2ac0-9bf6-11e8-84cb-519de54538b1.png">

6. The test will run, and Instruments will gather runtime profiling data about it. When it finishes, it should look like the screenshot below. Click in the **Points of Interest** instrument

    <img width="460" alt="screen shot 2018-08-09 at 4 52 15 pm" src="https://user-images.githubusercontent.com/1349578/43925823-6357b2de-9bf6-11e8-97ea-5ccaf05e78ac.png">

7. That will open up the list of "points of interest", but the test case only creates one region. Right-click on that region, and choose **Set Inspection Range and Zoom**

    <img width="453" alt="screen shot 2018-08-09 at 4 53 33 pm" src="https://user-images.githubusercontent.com/1349578/43925860-856aeda0-9bf6-11e8-82e9-4cbfd9269ce2.png">

8. What do you see? I see the screenshot below for a nearly 2 GB archived file, which looks like a constant amount of memory is being used the entire time. With what you're describing, I would expect it to look more like a ramp building the whole time. Let me know if you need further guidance

<img width="1396" alt="screen shot 2018-08-09 at 4 53 48 pm" src="https://user-images.githubusercontent.com/1349578/43925928-bc4cec1a-9bf6-11e8-9a7e-3dd88eddabd2.png">

**Comment by skito** [comment]
Well, better later than never. Sorry for the long delay.

Seems that I'm seeing the same numbers with profiling.
![screen shot 2018-10-02 at 14 21 52](https://user-images.githubusercontent.com/26840279/46345840-b7e83f80-c64e-11e8-8901-a0528e16887d.png)

However the numbers inside the debugger are different when running my app with the same archive.
![screen shot 2018-10-02 at 14 42 19](https://user-images.githubusercontent.com/26840279/46346747-c3893580-c651-11e8-9201-7cc4563205cb.png)

And it's significantly slowly. If I'm using the `extractFiles` the memory keeps under 30mb. 

Do you want me to pass the code I'm using? It's a one file independed class in swift. You just need to import it to your test project along with the UnrarKit framework and call the method?

**Comment by abbeycode** [comment]
@skito Sure, go ahead and send the swift file, but I think it would be even more useful if you could possibly send me a link to download the archive (DM me on Twitter @dovfrankel).

**Comment by skito** [comment]
Here's a link to download:
https://drive.google.com/drive/folders/18BhuHsF3lZF2rE-sWRYQHkcCBEKs_Iyw?usp=sharing

I'll keep it live for 24 hours. Some instructions below.

Example usage:
```swift
let extractor = Extractor(filePath: “source/path/file.rar“);
extractor.Extract(destinationPath: “destination/folder”, args: “1”);
```
In Extractor.swift there are two methods `BExtract` and `BExtract2` 
`BExtract` is using `extractBufferedData`
`BExtract2` is using `extractFiles`

If you want to switch between the methods just switch the names, e.g. rename `BExtract2` to `BExtract`.

Use the ...part1.rar file as path. Let me know for the result.

**Comment by abbeycode** [comment]
@skito I just got a chance to play around with the archive and code you sent. I'm glad to report that I was now able to reproduce what you're seeing. Further, I was able to reproduce it using my original test archive – by translating my test case to Swift. This appears to be caused (somehow) by bridging from Objective-C to Swift.

For a short-term workaround, you can write the code that calls `extractBufferedData` in Objective-C instead. I'll be investigating further to see what I can come up with.

**Comment by abbeycode** [comment]
I think I solved it! Can you try out the `swift-men-leak` branch? My tests showed it worked. Keeps memory usage around 12MB

**Comment by abbeycode** [comment]
I've merged the fix I introduced in PR #80, and am closing this issue. If you still experience this problem, please reopen the issue.
